### ISSUE TITLE
Temperature dependence of spin hall conductivity

### ISSUE THREAD

**Issue by ngohainc** [wannier-berri/wannier-berri#305]
Dear developers,

In some works, the temperature dependence of the SHC based on the Fermi-Dirac function has been presented. Could you tell us if we can do it by Wannierberri?

Thank you for the useful Wannierberri code.

**Comment by stepan-tsirkin** [comment]
In principle there is a parameter kBT in the `DynamicCalculator`

https://github.com/wannier-berri/wannier-berri/blob/3ee1fb0c4b3f751613fefac80f9ae71e074a034c/wannierberri/calculators/dynamic.py#L28-L29

which is temperature in eV. However that would mean that for every new temperature or chemical potential one needs to repeat the heavy Brillouin zone integration. A more efficient approach may be to evaluate the quantity at zero temperature for a range of Fermi levels once, and then for every other temperature smear if, by convoluting it with a Fermi-dirac smearing by using a relation

$$
f_{FD}(E_n-mu,T) = \int\limits_{-\infty}^{\infty}   f_{FD}(\varepsilon-\mu,T) \delta(\varepsilon-E_n) d\varepsilon  =  \int\limits_{-\infty}^{\infty}   -f'_{FD}(\varepsilon-\mu,T) \theta(\varepsilon-\mu) d\varepsilon 
$$

Where we used integration by parts $\theta'(x) = \delta(x)$. 

Thus, as far as $f_{FD}(\varepsilon-\mu,T=0) = \theta(\varepsilon-\mu) $, if you have a conductivity $\sigma(E_F,T=0)$, for any other temperature it is 

$$ 
\sigma(\mu,T) = \int\limits_{-\infty}^{\infty} \sigma(\varepsilon,T)\bigl (-f'_{FD}(\varepsilon-\mu)\bigr) d\varepsilon
$$

Such 1-dimensional integral may be done with any simple script.

**Comment by ngohainc** [comment]
Dear Prof. Stepan Tsirkin. Thank you for your kind interpretation. I truly appreciate it.

As you mentioned, 'kBT' is the parameter for the temperature dependence of SHC. However, I realize there is no change when setting at different temperatures. I tested on Pt using both Qiao and Ryoo methods with the input and result as in the attached picture. This result is totally different from a well-known PRL paper [PRL 100, 096401 (2008) https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.100.096401] where the SHC of Pt drastically reduces from 2200(ℏ/e)(Ω cm)−1 to 240 (ℏ/e)(Ω cm)−1 as T rises from 0 to 300 K.

Also, when testing the Optical conductivity of Fe,  there seems no clear difference between the cases of 0K and 300K.
Could you tell us the possible reason behind these abnormalities?

![image](https://github.com/wannier-berri/wannier-berri/assets/107199859/48400002-ba5f-496e-815f-65075b58efbc)
![image](https://github.com/wannier-berri/wannier-berri/assets/107199859/2f8bdb40-d9bf-4101-a86b-2095892dfc27)

**Comment by stepan-tsirkin** [comment]
@ngohainc ,

Your results looks like smth that makes sense to me, but I do not understand how that PRL obtained such a strong temperature dependence. 

Your curve for PT looks consistent with their low-temp curve. The peak at the Fermi level is wide (~0.5 eV half width at half height), much wider than the 300K value. If temperature is included only in the Fermi factors, the 300k cannot affect much. In the paper I do not see anything else on temperature effects ( like lattice constant, or phonon-mediated bandstructure change or anything like that). Maybe it is worth contacting the Author of that PRL for more information. If you get such, please keep us updated.

@manxkim , @jaemolihm did you ever study that paper in detail?

**Comment by jaemolihm** [comment]
I had the very same question about that paper, but I do not have an answer either...

**Comment by ngohainc** [comment]
Many thanks, Prof. Stepan Tsirkin. I have emailed the corresponding Author of the PRL to learn more

**Comment by jhryoo** [comment]
Although I haven't thoroughly studied the temperature dependence of SHC or AHC, I believe that your results are more plausible than that of PRL 100, 096401 (2008), at least within the formalism of the Kubo formula.

In their paper, Guo mentioned that the strong temperature dependence of SHC is due to the near-degeneracies near the Fermi energy at X and L points. According to their paper, in Eq. (1), the only part the temperature dependence comes in is the Fermi-Dirac occupancy factor. However, the energy differences between unoccupied and occupied states at X and L points are a few hundred meVs. Given that these energies are not so small compared to the room-temperature energy scale (kT ~ 0.025 eV), the origin of the strong temperature dependence Guo observed is not entirely clear to me. If their calculation is indeed correct, I guess that the temperature dependence is due to the large contribution to SHC from points slightly far from X point (See Fig. 2a in Guo's paper; there is a large positive peak next to X point), which may be easily suppressed by the increase in the temperature, since in that region the energy of the unoccupied states is much closer to the Fermi energy. Nevertheless, if this scenario was true, I think this effect must be observed in calculations using Wannier-Berri as well.

There is another paper that studied the temperature dependence of SHC of Pt from the first principles: PRL 117, 167204 (2016), using a somewhat different method. They used the Kubo-Bastin equation to calculate the extrinsic as well as the intrinsic SHC and also considered the effect of impurity and thermally-induced lattice vibration using CPA. According to their supplemental material (https://journals.aps.org/prl/supplemental/10.1103/PhysRevLett.117.167204/supplement_Last.pdf), the authors has found that as the temperature is increased from 0K to 300K the intrinsic SHC of Pt is decreased by less than 20% (see Fig. S6 in the above link) rather than by a factor of 10 as Guo observed. Even this 20% decrease may originate from the thermal vibration of the lattice (band distortion and finite transport lifetime of electrons), rather than the change in Fermi-Dirac occupancy factor.

**Comment by ngohainc** [comment]
Dear Developers,

In the dynamic.py and __utility.py I can see FermiDirac has been given in the Kubo-greenwood formula. In the same approach, I'm trying to implement FermiDirac in the static AHC because the dynamic calculation consumes much more RAM space than static calculation. I guess the Omega (in covariant.py) and StaticCalculator (in static.py) should be the places to insert more code. However, since I am an experimentalist without experience in coding I have not yet found the correct way to implement it.

Could you tell me the hint for such an idea?

**Comment by stepan-tsirkin** [comment]
@ngohainc , sorry, I missed your message, so did not reply for such long time. Is your question still of your interest?

If so, what integral are you willing to compute? If it is AHC for finite temperature, you do not need to implement anything new. Just integrate at zero temperature, and then apply F-D smearing to the result.

**Comment by ngohainc** [comment]
Dear Prof. Stepan, 
Thank you for your kind reply. Yes, I'm still interested in the topic of evaluating AHC at finite temperatures.

Can you interpret via a certain case of Fe (bcc) for instance? As you said, we first integrate at zero temperature to obtain σ(Ef, T=0). Then, could you tell us the additional script for using the F-D smearing to obtain AHC at a finite temperature (300K)?

**Comment by stepan-tsirkin** [comment]
Here is an example for DOS, but works for AHC in a similar way:

```python

calculators = {"dos": wberri.calculators.static.DOS(Efermi = np.linspace(6.0,6.1,201),
        tetra=False # intentionally set to False to show the effect of smoothing. For good quality results, set it to True
        ) }

grid = wberri.Grid(system, length=200, NKFFT=[6 ,6,8] )   

result = wberri.run(system,
            grid=grid,
            calculators = calculators,
            parallel=parallel,
            adpt_num_iter=0,
            use_irred_kpt=True,
            symmetrize=True,
            print_Kpoints=False,
            )

res = result.results["dos"]
Efermi_eV = res.Energies[0]
Efermi_meV = Efermi_eV*1000-6100

plt.plot(Efermi_meV,res.data, label = "no smoother")
plt.plot(Efermi_meV,res.dataSmooth, label="default smoother")

y={}
for T in [1,10,50,100]:
    res.set_smoother(wberri.smoother.FermiDiracSmoother( E=Efermi_eV, T_Kelvin=T))
    del res.dataSmooth # this is important, because dataSmooth is a cached property
    y[T] = res.dataSmooth
    plt.plot(Efermi_meV,y[T], label=f"Fermi-Dirac smoother T={T}K")

plt.legend()
plt.xlabel("$E-E_F$ [meV]")
plt.ylabel("DOS [1/eV]")
plt.show()

```

![image](https://github.com/wannier-berri/wannier-berri/assets/22172977/f2347328-d957-46fb-9882-254a5bbb5da6)

**Comment by ngohainc** [comment]
Thank you very much for your kind support, Developers.

**Comment by ngohainc** [comment]
When running the testing code you provided I saw "AttributeError: can't delete attribute". As I understand, the error is at this line:
            del res.dataSmooth  # this is important, because dataSmooth is a cached property
Could you tell me the way to fix this error, Developer?

**Comment by ngohainc** [comment]
Yes, by using the additional code you provide it works now. I test the AHC for Fe from your resource (wannier-berri/tests/data//Fe_Wannier90). However, I found out that there is no difference between cases: no smoother, default smoother, 1K, 10K, 50K, 300K. I checked the numeric data and saw that all the corresponding data are exactly the same. Can you what causes this issue? Here are the code, graph, and numeric data.

system=wberri.System_w90(seedname='Fe',berry=True)
generators = ['Inversion','C4z','TimeReversal*C2x']
system.set_symmetry(generators)
calculators = {"ahc": wberri.calculators.static.AHC(Efermi = np.linspace(17.125,18.125,201),
        tetra=False ) }
grid = wberri.Grid(system, length=200, NKFFT=[8 ,8,8] )   
parallel=wberri.Parallel()

result = wberri.run(system,
            grid=grid,
            calculators = calculators,
            parallel=parallel,
            adpt_num_iter=2,
            use_irred_kpt=True,
            symmetrize=True,
            print_Kpoints=False,
            )

res = result.results["ahc"]
Efermi_eV = res.Energies[0]
Efermi_meV = Efermi_eV*1000-17625

plt.plot(Efermi_meV,res.data[:,2], label = "no smoother")
plt.plot(Efermi_meV,res.dataSmooth[:,2], label="default smoother")

y={}
for T in [1,10,50,100]:
    res.set_smoother(wberri.smoother.FermiDiracSmoother( E=Efermi_eV, T_Kelvin=T))
    try:
      del res.dataSmooth
    except AttributeError:
      pass
    y[T] = res.dataSmooth[:,2]
    plt.plot(Efermi_meV,y[T], label=f"Fermi-Dirac smoother T={T}K")

plt.legend()
plt.xlabel("$E-E_F$ [meV]")
plt.ylabel(r'$\sigma_{xy}^{\rm AHC}$ $10^{-2}$ (Ω.cm)$^{-1}$')

![AHC-mod01](https://github.com/user-attachments/assets/65f47ff5-4568-4216-87e2-9f91c5d75d2a)
[DATA.txt](https://github.com/user-attachments/files/16490342/DATA.txt)

**Comment by stepan-tsirkin** [comment]
Hi. With your same code I get the following, both with 0.17.0 (latest in pip) and the master branch.

![Fe-ahc-Tdep-2](https://github.com/user-attachments/assets/938d9eed-5224-46f9-ac9f-2f733cb1e8f4)

Also. I do not get the error when deleting the dataSmooth. Which version are you using? probably, it is an older version that was using lazy_property instead of cached_property, or smth else

**Comment by ngohainc** [comment]
Yes, you are right. I use the older version 0.14.1, thus, I got that kind of error. I have installed the latest version 0.17.0 and got the same result as yours. 
Many thanks, Developer for your very kind support.

**Comment by ngohainc** [comment]
I tested for SHC on the newest version (0.17.0) but it does not work.  In the latest version, it seems that I need to write some input codes in a different style. Could you tell me the error I made in this bellow code?

system=wberri.System_w90(seedname='Fe',berry=True)
generators = ['Inversion','C4z','TimeReversal*C2x']
system.set_symmetry(generators)

calculators = {"shc": wberri.calculators.static.SHC(kwargs_formula={'spin_current_type':'qiao'}, Efermi = np.linspace(17.125,18.125,2001),
        tetra=False ) }
grid = wberri.Grid(system, length=200, NKFFT=[8 ,8,8] )   
parallel=wberri.Parallel()

result = wberri.run(system,
            grid=grid,
            calculators = calculators,
            parallel=parallel,
            adpt_num_iter=3,
            use_irred_kpt=True,
            symmetrize=True,
            print_Kpoints=False,
            )

res = result.results["shc"]
Efermi_eV = res.Energies[0]
Efermi_meV = Efermi_eV*1000-17625

plt.plot(Efermi_meV,res.data[:,2], label = "no smoother")
plt.plot(Efermi_meV,res.dataSmooth[:,2], label="default smoother")

y={}
for T in [1,10,50,100]:  
    res.set_smoother(wberri.smoother.FermiDiracSmoother( E=Efermi_eV, T_Kelvin=T))
    del res.dataSmooth
    y[T] = res.dataSmooth[:,2]
    plt.plot(Efermi_meV,y[T], label=f"Fermi-Dirac smoother T={T}K")

plt.legend()
plt.xlabel("$E-E_F$ [meV]")
plt.ylabel(r'$\sigma_{xy}^{\rm SHC}$ $10^{-2}$ (Ω.cm)$^{-1}$')

**Comment by ngohainc** [comment]
Yes, when adding SHCqiao=True in System_w90, I still face other errors. It says that:
 AttributeError: 'numpy.ndarray' object has no attribute 'set_smoother'
As I know, it comes from this line
    res.set_smoother(wberri.smoother.FermiDiracSmoother( E=Efermi_eV, T_Kelvin=T))

Here is the code that I tested on.

system=wberri.System_w90(seedname='Fe',berry=True,SHCqiao=True)
generators = ['Inversion','C4z','TimeReversal*C2x']
system.set_symmetry(generators)

calculators = {"shc": wberri.calculators.static.SHC(kwargs_formula={'spin_current_type':'qiao'}, Efermi = np.linspace(17.125,18.125,2001),
        tetra=False ) }
grid = wberri.Grid(system, length=200, NKFFT=[8 ,8,8] )
Efermiev = np.linspace(17.125,18.125,2001)
parallel=wberri.Parallel()

result = wberri.run(system,
            grid=grid,
            calculators = calculators,
            parallel=parallel,
            adpt_num_iter=3,
            use_irred_kpt=True,
            symmetrize=True,
            print_Kpoints=False,
            )

res = result.results["shc"].data[:, 0, 1, 2]
resmooth = result.results["shc"].dataSmooth[:, 0, 1, 2]
Efermi_meV = Efermiev*1000-17625

plt.plot(Efermi_meV,res, label = "no smoother")
plt.plot(Efermi_meV,resmooth, label="default smoother")

y={}
for T in [1,10,50,100]: 
    res.set_smoother(wberri.smoother.FermiDiracSmoother( E=Efermi_eV, T_Kelvin=T))
    del res.dataSmooth
    y[T] = res.dataSmooth[:, 0, 1, 2]
    plt.plot(Efermi_meV,y[T], label=f"Fermi-Dirac smoother T={T}K")

plt.legend()
plt.xlabel("$E-E_F$ [meV]")
plt.ylabel(r'$\sigma_{xy}^{\rm AHC}$ $10^{-2}$ (Ω.cm)$^{-1}$')
plt.savefig('shc.png',bbox_inches='tight',dpi=1000)

**Comment by stepan-tsirkin** [comment]
```python 
res = result.results["shc"].data
``` 

is a numpy array, you cannot set a smoother to it. Smth like this should work:

```python
res = result.results["shc"]
res.set_smoother(...)
res.dataSmooth[:,0,1,2]
```



Also, could you format your messages to highlight the syntax, see here

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Comment by ngohainc** [comment]
Yes, it runs well now. Thank you for letting me know how to highlight the syntax.
